eyecite/resolve.py

```diff
@@ -0,0 +1,180 @@
+    citation: ShortCaseCitation,
+    resolved_full_cites: Dict[ResourceType, FullCaseCitation],
+) -> Optional[ResourceType]:
+    """
+    Try to match shortcase citations by checking whether their reporter and
+    volume number matches those of any of the previously resolved full
+    citations. If there are multiple possible matches, try to refine by also
+    checking whether their antecedent_guess appears in either the defendant
+    or plaintiff field of any of the previously resolved full citations.
+    """
+    candidates: List[Tuple[ResourceType, FullCaseCitation]] = []
+    for resource, full_citation in resolved_full_cites.items():
+        if (
+            citation.reporter == full_citation.reporter
+            and citation.volume == full_citation.volume
+        ):
+            # Append both keys and values for further refinement below
+            candidates.append((resource, full_citation))
+
+    # Only accept if one candidate remains
+    if len(candidates) == 1:
+        return candidates[0][0]
+
+    # Otherwise, if there is an antecedent guess, try to refine further
+    elif citation.antecedent_guess:
+        refined_candidates: List[ResourceType] = []
+        ag: str = strip_punct(citation.antecedent_guess)
+
+        for resource, full_citation in candidates:
+            if full_citation.defendant and ag in full_citation.defendant:
+                refined_candidates.append(resource)
+            elif full_citation.plaintiff and ag in full_citation.plaintiff:
+                refined_candidates.append(resource)
+
+        # Only accept if one candidate remains
+        return refined_candidates[0] if len(refined_candidates) == 1 else None
+
+    # Otherwise, nothing left to try
+    else:
+        return None
+
+
+def _resolve_supra_citation(
+    citation: SupraCitation,
+    resolved_full_cites: Dict[ResourceType, FullCaseCitation],
+) -> Optional[ResourceType]:
+    """
+    Try to resolve supra citations by checking whether their antecedent_guess
+    appears in either the defendant or plaintiff field of any of the
+    previously resolved full citations.
+    """
+    if not citation.antecedent_guess:  # If no guess, can't do anything
+        return None
+
+    candidates: List[ResourceType] = []
+    ag: str = strip_punct(citation.antecedent_guess)
+
+    for resource, full_citation in resolved_full_cites.items():
+        if full_citation.defendant and ag in full_citation.defendant:
+            candidates.append(resource)
+        elif full_citation.plaintiff and ag in full_citation.plaintiff:
+            candidates.append(resource)
+
+    # Only accept if one candidate remains
+    return candidates[0] if len(candidates) == 1 else None
+
+
+def _resolve_id_citation(
+    citation: IdCitation,
+    last_resolution: ResourceType,
+) -> Optional[ResourceType]:
+    """
+    Resolve id citations to the resource of the previously resolved
+    citation.
+    """
+    return last_resolution
```

Comment:
Does this need a check that `last_resolution` is not `None`?

Reply:
I don't think so, because if `last_resolution` is `None`, then we would return `None` anyway. Thus, we can simply return `last_resolution` itself, no matter what.